src/scheduler/queue.c

```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright (C) 1994-2019 Altair Engineering, Inc.
+ * For more information, contact Altair at www.altair.com.
+ *
+ * This file is part of the PBS Professional ("PBS Pro") software.
+ *
+ * Open Source License Information:
+ *
+ * PBS Pro is free software. You can redistribute it and/or modify it under the
+ * terms of the GNU Affero General Public License as published by the Free
+ * Software Foundation, either version 3 of the License, or (at your option) any
+ * later version.
+ *
+ * PBS Pro is distributed in the hope that it will be useful, but WITHOUT ANY
+ * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE.
+ * See the GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Commercial License Information:
+ *
+ * For a copy of the commercial license terms and conditions,
+ * go to: (http://www.pbspro.com/UserArea/agreement.html)
+ * or contact the Altair Legal Department.
+ *
+ * Altair’s dual-license business model allows companies, individuals, and
+ * organizations to create proprietary derivative works of PBS Pro and
+ * distribute them - whether embedded or bundled with other software -
+ * under a commercial license agreement.
+ *
+ * Use of Altair’s trademarks, including but not limited to "PBS™",
+ * "PBS Professional®", and "PBS Pro™" and Altair’s logos is subject to Altair's
+ * trademark licensing policies.
+ *
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <errno.h>
+
+#include "constant.h"
+#include "log.h"
+#include "queue.h"
+
+
+/**
+ * @brief	Constructor for the data structure 'queue'
+ *
+ * @param	void
+ *
+ * @return queue *
+ * @retval a newly allocated queue object
+ * @retval NULL for malloc error
+ */
+ds_queue *
+new_ds_queue(void)
+{
+	ds_queue *ret_obj = NULL;
+
+	ret_obj = malloc(sizeof(ds_queue));
+	if (ret_obj == NULL) {
+		log_err(errno, __func__, MEM_ERR_MSG);
+		return NULL;
+	}
+
+	ret_obj->min_size = QUEUE_DS_MIN_SIZE;
+	ret_obj->content_arr = NULL;
+	ret_obj->front = 0;
+	ret_obj->rear = 0;
+	ret_obj->q_size = 0;
+
+	return ret_obj;
+}
+
+/**
+ * @brief	Destructor for a queue object
+ *
+ * @param[in]	obj - a queue object to deallocate
+ *
+ * @return void
+ */
+void
+free_ds_queue(ds_queue *obj)
```

Comment:
can you rename obj to queue because everywhere else you are using obj as the object to store